File: sample/src/main/java/com/urbanairship/sample/SampleAsyncLiveUpdate.java
package com.urbanairship.sample;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.RemoteViews;

import com.bumptech.glide.Glide;
import com.urbanairship.json.JsonMap;
import com.urbanairship.liveupdate.CallbackLiveUpdateNotificationHandler;
import com.urbanairship.liveupdate.LiveUpdate;
import com.urbanairship.liveupdate.LiveUpdateEvent;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.core.app.NotificationCompat;

/**
 * Sample sports live update handler, with support for asynchronously fetching team images during
 * onUpdate.
 */
public class SampleAsyncLiveUpdate implements CallbackLiveUpdateNotificationHandler {

    private final ExecutorService backgroundExecutor = Executors.newSingleThreadExecutor();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    @Override
    public void onUpdate(@NonNull Context context,
                         @NonNull LiveUpdateEvent event,
                         @NonNull LiveUpdate update,
                         @NonNull LiveUpdateResultCallback callback) {

        Log.d("SampleAsyncLiveUpdate", "onUpdate: action=" + event + ", update=" + update);

        if (event == LiveUpdateEvent.END) {
            // Dismiss the live update on END. The default behavior will leave the Live Update
            // in the notification tray until the dismissal time is reached or the user dismisses it.
            callback.cancel();
            return;
        }

        JsonMap content = update.getContent();
        int teamOneScore = content.opt("team_one_score").getInt(0);
        int teamTwoScore = content.opt("team_two_score").getInt(0);
        String teamOneName = content.opt("team_one_name").getString("Foxes");
        String teamTwoName = content.opt("team_two_name").getString("Tigers");
        String teamOneImageUrl = content.opt("team_one_image").getString("");
        String teamTwoImageUrl = content.opt("team_two_image").getString("");
        String statusUpdate = content.opt("status_update").optString();

        RemoteViews bigLayout = new RemoteViews(context.getPackageName(), R.layout.live_update_notification_big);
        bigLayout.setTextViewText(R.id.teamOneName, teamOneName);
        bigLayout.setTextViewText(R.id.teamTwoName, teamTwoName);
        bigLayout.setTextViewText(R.id.teamOneScore, String.valueOf(teamOneScore));
        bigLayout.setTextViewText(R.id.teamTwoScore, String.valueOf(teamTwoScore));
        bigLayout.setTextViewText(R.id.statusUpdate, statusUpdate);

        RemoteViews smallLayout = new RemoteViews(context.getPackageName(), R.layout.live_update_notification_small);
        smallLayout.setTextViewText(R.id.teamOneScore, String.valueOf(teamOneScore));
        smallLayout.setTextViewText(R.id.teamTwoScore, String.valueOf(teamTwoScore));

        Intent launchIntent = context.getPackageManager()
                                     .getLaunchIntentForPackage(context.getPackageName())
                                     .addCategory(update.getName())
                                     .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_SINGLE_TOP)
                                     .setPackage(null);

        PendingIntent contentIntent = PendingIntent.getActivity(
                context, 0, launchIntent, PendingIntent.FLAG_IMMUTABLE);

        NotificationCompat.Builder builder =
                new NotificationCompat.Builder(context, "sports")
                        .setSmallIcon(R.drawable.ic_notification)
                        .setPriority(NotificationCompat.PRIORITY_MAX)
                        .setCategory(NotificationCompat.CATEGORY_EVENT)
                        .setStyle(new NotificationCompat.DecoratedCustomViewStyle())
                        .setCustomContentView(smallLayout)
                        .setCustomBigContentView(bigLayout)
                        .setContentIntent(contentIntent);

        if (teamOneImageUrl.isEmpty() && teamTwoImageUrl.isEmpty()) {
            callback.ok(builder);
        } else {
            fetchTeamIcons(context, teamOneImageUrl, teamTwoImageUrl, (teamOneBmp, teamTwoBmp) -> {
                if (teamOneBmp != null && teamTwoBmp != null) {
                    smallLayout.setImageViewBitmap(R.id.teamTwoImage, teamTwoBmp);
                    smallLayout.setImageViewBitmap(R.id.teamOneImage, teamOneBmp);
                    bigLayout.setImageViewBitmap(R.id.teamOneImage, teamOneBmp);
                    bigLayout.setImageViewBitmap(R.id.teamTwoImage, teamTwoBmp);
                }

                NotificationResult result = callback.ok(builder);

                // ...
                result.getNotification();
                result.getNotificationId();
                result.getNotificationTag();
            });
        }
    }

    private void fetchTeamIcons(Context context, String teamOneUrl, String teamTwoUrl, TeamIconsCallback callback) {
        backgroundExecutor.execute(() -> {
            try {
                Bitmap teamOneBitmap = fetchBitmap(context, teamOneUrl);
                Bitmap teamTwoBitmap = fetchBitmap(context, teamTwoUrl);

                mainHandler.post(() -> callback.onIconsReady(teamOneBitmap, teamTwoBitmap));
            } catch (Exception e) {
                mainHandler.post(() -> callback.onIconsReady(null, null));
            }
        });
    }

    @WorkerThread
    private Bitmap fetchBitmap(Context context, String url) throws ExecutionException, InterruptedException {
        return Glide.with(context)
                    .asBitmap()
                    .load(url)
                    .submit()
                    .get();
    }

    private interface TeamIconsCallback {
        @MainThread
        void onIconsReady(@Nullable Bitmap teamOneBitmap, @Nullable Bitmap teamTwoBitmap);
    }
}

File: devapp/src/main/java/com/urbanairship/devapp/SampleAsyncLiveUpdate.kt
package com.urbanairship.devapp

import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.util.Log
import android.widget.RemoteViews
import androidx.annotation.WorkerThread
import androidx.core.app.NotificationCompat
import com.urbanairship.AirshipDispatchers
import com.urbanairship.json.JsonMap
import com.urbanairship.liveupdate.CallbackLiveUpdateNotificationHandler
import com.urbanairship.liveupdate.LiveUpdate
import com.urbanairship.liveupdate.LiveUpdateEvent
import java.util.concurrent.ExecutionException
import com.bumptech.glide.Glide
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Sample sports live update handler, with support for asynchronously fetching team images during
 * onUpdate.
 */
class SampleAsyncLiveUpdate(
    backgroundDispatcher: CoroutineDispatcher = AirshipDispatchers.newSerialDispatcher(),
) : CallbackLiveUpdateNotificationHandler {

    private val scope = CoroutineScope(backgroundDispatcher + SupervisorJob())

    override fun onUpdate(
        context: Context,
        event: LiveUpdateEvent,
        update: LiveUpdate,
        resultCallback: CallbackLiveUpdateNotificationHandler.LiveUpdateResultCallback
    ) {
        Log.d("SampleAsyncLiveUpdate", "onUpdate: action=$event, update=$update")

        if (event == LiveUpdateEvent.END) {
            // Dismiss the live update on END. The default behavior will leave the Live Update
            // in the notification tray until the dismissal time is reached or the user dismisses it.
            resultCallback.cancel()
            return
        }

        val contentUpdate = ContentUpdate.fromJson(update.content)

        val bigLayout = RemoteViews(context.packageName, R.layout.live_update_notification_big)
        contentUpdate.fillLayout(bigLayout, layoutType = ContentUpdate.LayoutType.BIG)

        val smallLayout = RemoteViews(context.packageName, R.layout.live_update_notification_small)
        contentUpdate.fillLayout(smallLayout, layoutType = ContentUpdate.LayoutType.SMALL)

        val launchIntent = context.packageManager
            .getLaunchIntentForPackage(context.packageName)
            ?.addCategory(update.name)
            ?.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_SINGLE_TOP)
            ?.setPackage(null)

        val contentIntent = PendingIntent.getActivity(
            context, 0, launchIntent, PendingIntent.FLAG_IMMUTABLE
        )

        val builder = NotificationCompat.Builder(context, "sports")
            .setSmallIcon(R.drawable.ic_notification)
            .setPriority(NotificationCompat.PRIORITY_MAX)
            .setCategory(NotificationCompat.CATEGORY_EVENT)
            .setStyle(NotificationCompat.DecoratedCustomViewStyle())
            .setCustomContentView(smallLayout).setCustomBigContentView(bigLayout)
            .setContentIntent(contentIntent)

        scope.launch {
            val icons = contentUpdate.loadTeamIcons(context)

            withContext(Dispatchers.Main) {
                if (icons.teamOne != null && icons.teamTwo != null) {
                    smallLayout.setImageViewBitmap(R.id.teamTwoImage, icons.teamTwo)
                    smallLayout.setImageViewBitmap(R.id.teamOneImage, icons.teamOne)
                    bigLayout.setImageViewBitmap(R.id.teamOneImage, icons.teamOne)
                    bigLayout.setImageViewBitmap(R.id.teamTwoImage, icons.teamTwo)
                }

                resultCallback.ok(builder)
            }
        }
    }

    private data class ContentUpdate(
        val teamOne: Team,
        val teamTwo: Team,
        val statusUpdate: String
    ) {
        data class Team (
            val name: String,
            val score: Int,
            val imageUrl: String
        ) {
            @WorkerThread
            @Throws(ExecutionException::class, InterruptedException::class)
            fun fetchBitmap(context: Context): Bitmap? {
                if (imageUrl.isEmpty()) {
                    return null
                }

                return Glide.with(context).asBitmap().load(imageUrl).submit().get()
            }
        }

        enum class LayoutType {
            BIG, SMALL
        }

        fun fillLayout(layout: RemoteViews, layoutType: LayoutType) {
            layout.setTextViewText(R.id.teamOneScore, teamOne.score.toString())
            layout.setTextViewText(R.id.teamTwoScore, teamTwo.score.toString())

            when(layoutType) {
                LayoutType.BIG -> {
                    layout.setTextViewText(R.id.teamOneName, teamOne.name)
                    layout.setTextViewText(R.id.teamTwoName, teamTwo.name)
                    layout.setTextViewText(R.id.statusUpdate, statusUpdate)
                }
                else -> {}
            }
        }

        @WorkerThread
        fun loadTeamIcons(context: Context): TeamLogo {
            val result = listOf(teamOne, teamTwo).map {
                try {
                    it.fetchBitmap(context)
                } catch (_: Exception) {
                    null
                }
            }.let {
                TeamLogo(it[0], it[1])
            }

            return result
        }

        data class TeamLogo(
            val teamOne: Bitmap?,
            val teamTwo: Bitmap?
        )

        companion object {
            private const val TEAM_ONE_PREFIX = "team_one_"
            private const val TEAM_TWO_PREFIX = "team_two_"

            private const val SCORE_KEY = "score"
            private const val NAME_KEY = "name"
            private const val IMAGE_KEY = "image"

            private const val STATUS_UPDATE_KEY = "status_update"

            fun fromJson(value: JsonMap): ContentUpdate {
                return ContentUpdate(
                    teamOne = Team(
                        name = value.opt(TEAM_ONE_PREFIX + NAME_KEY).getString("Foxes"),
                        score = value.opt(TEAM_ONE_PREFIX + SCORE_KEY).getInt(0),
                        imageUrl = value.opt(TEAM_ONE_PREFIX + IMAGE_KEY).getString("")
                    ),
                    teamTwo = Team(
                        name = value.opt(TEAM_TWO_PREFIX + NAME_KEY).getString("Tigers"),
                        score = value.opt(TEAM_TWO_PREFIX + SCORE_KEY).getInt(0),
                        imageUrl = value.opt(TEAM_TWO_PREFIX + IMAGE_KEY).getString("")
                    ),
                    statusUpdate = value.opt(STATUS_UPDATE_KEY).optString()
                )
            }
        }
    }
}
